lib/src/api/servicer.rs

```diff
@@ -0,0 +1,101 @@
+use crate::api::from_proto;
+use crate::object_id::ObjectId;
+use crate::repo::ReadonlyRepo;
+use crate::settings::UserSettings;
+use config::Config;
+use itertools::Itertools;
+use jj_api::objects::{Change as ChangeProto, Workspace as WorkspaceProto};
+use jj_api::rpc::{ListWorkspacesRequest, ListWorkspacesResponse};
+use jj_lib::op_store::OperationId;
+use jj_lib::operation::Operation;
+use jj_lib::repo::RepoLoader;
+use jj_lib::workspace::WorkspaceLoader;
+
+use std::sync::Arc;
+use tonic::Status;
+
+/// The servicer handles all requests going to jj-lib. Eventually, ideally, jj-cli
+/// will interact with jj-lib purely through this class.
+pub struct Servicer {
+    default_workspace_loader: Option<WorkspaceLoader>,
+    user_settings: UserSettings,
+}
+
+impl Servicer {
+    pub fn new(default_workspace_loader: Option<WorkspaceLoader>) -> Self {
+        Self {
+            default_workspace_loader,
+            user_settings: UserSettings::from_config(Config::default()),
+        }
+    }
+
+    fn workspace_loader(
+        &self,
+        opts: &Option<jj_api::objects::RepoOptions>,
+    ) -> Result<WorkspaceLoader, Status> {
+        opts.as_ref()
+            .map(|opts| from_proto::path(&opts.repo_path))
```

Comment:
This is a lot of work to just get one field, which is why I'm considering a mechanism where we just have:
`grpc_servicer.rs` would write:
```
// servicer.rs
fn list_workspaces(TypeSafeListWorkspacesRequest) -> Result<TypeSafeListWorkspacesResponse, Status> {
    ...
}

// grpc_servicer.rs

async fn list_workspaces(
        &self,
        request: Request<ListWorkspacesRequest>,
    ) -> Result<Response<ListWorkspacesResponse>, Status> {
        self.servicer
            .list_workspaces(request.get_ref().try_into()?)
            .to_proto()
            .map(Response::new)
    }
```